classes/merge_request.php

```diff
@@ -0,0 +1,150 @@
+     * could not end with success).
+     * @var integer
+     */
+    const COMPLETED_WITH_ERRORS = 7;
+    /**
+     * Reference table for merge requests.
+     */
+    const TABLE_MERGE_REQUEST = 'tool_mergeusers_queue';
+     /**
+     * Reference table for merge requests. Before web service integration.
+     */
+    const TABLE_MERGE_REQUEST_OLD = 'tool_mergeusers';
+    /**
+     * Reference table for users.
+     */
+    const TABLE_USERS = 'user';
+    
+    private $data;
+
+    private function __construct(stdClass $data) {
+        $this->data = $data;
+    }
+
+    public static function from(stdClass $data) {
+        return new self($data);
+    }
+
+    public function __get($name)
+    {
+        if (isset($this->data->${name})) {
+            $value = $this->data->${name};
+            if ($name == 'log') {
+                $value = json_decode($value, true);
+            }
+            return $value;
+        }
+        return null;
+    }
+
+    public function __set($name, $value)
+    {
+        if (isset($this->data->${name})) {
+            if ($name == 'log') {
+                $value = json_encode($value);
+            }
+            $this->data->${name} = $value;
+        }
+    }
+
+    public function convert_log_to_new_format() {
+        if (!isset($this->data->log) || empty($this->data->log)) {
+            return;
+        }
+        $logs = [];
+        $logs[1] = json_decode($this->data->log, true);
+        $this->data->log = json_encode($logs);
+    }
+
+    public function get_record(): \stdClass {
+        return $this->data;
+    }
+
+    public function export_data_to_new_table(string $oldtable, 
+                                            string $newtable) {
+        global $DB;
+        $filter = array('status' => self::QUEUED_NOT_PROCESSED);
+        $limitfrom=0;
+        $limitnum=0; 
+        $sort = "id DESC";
+        $records = $DB->get_records($oldtable, $filter, $sort, 
+                                'id, touserid, fromuserid, success, timemodified, log', 
+                                $limitfrom, $limitnum);
+        if (!$records) {
+            return $records;
+        }
+        foreach ($records as $item) {
+            //insert item into new table
+            if ($item->status == 1) {
+                $status = merge_request::COMPLETED_WITH_SUCCESS;
+            } else if ($item->status == 0) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+            }
+        $logs = [];
+        $logs[1] = json_decode($item->log, true);
+       
+            $idrecord = $DB->insert_record(
+                merge_request::TABLE_MERGE_REQUEST ,
+                [
+                    'removeuserid' => $item->fromuserid,  
+                    'keepuserid' => $item->touserid,
+                    'timeadded' => $timemodified,
+                    'timemodified' => time(),
+                    'status' => $status,
+                    'log' =>  json_encode($logs)
```

Comment:
it's no so clear to me as proceed.

Reply:
I propose you to do something like this:

```php
public function export_data_to_new_table() {
        global $DB;
        $filter = array();
        $records = $DB->get_records(self::TABLE_MERGE_REQUEST_OLD, ['status' => self::QUEUED_NOT_PROCESSED], 'id ASC', 
                                'id, touserid, fromuserid, success, timemodified, log');
        if (!$records) {
            // There is no record to migrate. That's all!
            return;
        }

        $oldrequests = [];
        $orderedoldrequests = [];
        foreach ($records as $item) {
            if (!isset($oldrequests[$item->fromuserid])) {
               $oldrequests[$item->fromuserid] = [];
            }
            if (isset($oldrequests[$item->fromuserid][$item->touserid)) {
                 $baseitem = $oldrequests[$item->fromuserid][$item->touserid];
            } else {
                $baseitem = new \stdClass();
                $baseitem->removeuserfield = 'id';
                $baseitem->removeuservalue = $item->fromuserid;
                $baseitem->removeuserid = $item->fromuserid;
                $baseitem->keepuserfield = 'id';
                $baseitem->keepuservalue = $item->touserid;
                $baseitem->keepuserid = $item->touserid;
                $baseitem->timeadded = $item->timemodified;
                $baseitem->log = [];
                $oldrequests[$item->fromuserid][$item->touserid] = $baseitem;
                $orderedoldrequests[$baseitem->timeadded] = $baseitem;
            }

            // Old logs are obtained in ASC order, so we can safely update this values.
            $baseitem->timemodified = $item->timemodified;
            $baseitem->timecompleted = $item->timemodified;
            // Append logs to the list.
            $baseitem->log[$item->timemodified] =  json_decode($item->log, true);
            $baseitem->status = ($item->status == 1) ? merge_request::COMPLETED_WITH_SUCCESS: 
 merge_request::COMPLETED_WITH_ERRORS;
       }

       // Insert ordered and simplified old records into new format.
       foreach ($orderedoldrecords as $newrecord) {
           $newrecord->log = json_encode($newrecord->log);
           $DB->insert_record(merge_merge_request::TABLE_MERGE_REQUEST, $newrecord);
        }
    }
```

Even though that I used in this example the stdClass instance, I think it would be rather simplified using the merge_request class here. But it needs a bit of care to change it.

As in other ocasions, this code is an example and written directly here in the comment. It is not tested. Just provided to you to help you.

Reply:
Let me know if you need something else to this respect.

Reply:
done